src/librustc_typeck/check/mod.rs

```diff
@@ -3752,14 +3752,35 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         if let Some(ref init) = local.init {
             let init_ty = self.check_decl_initializer(local, &init);
             if init_ty.references_error() {
+                // Override the types everywhere with `types.err` to avoid knock down errors.
                 self.write_ty(local.hir_id, init_ty);
+                self.write_ty(local.pat.hir_id, init_ty);
+                self.locals.borrow_mut().insert(local.hir_id, LocalTy {
+                    decl_ty: t,
+                    revealed_ty: init_ty,
+                });
+                self.locals.borrow_mut().insert(local.pat.hir_id, LocalTy {
+                    decl_ty: t,
+                    revealed_ty: init_ty,
+                });
             }
         }
 
         self.check_pat_top(&local.pat, t, None);
         let pat_ty = self.node_ty(local.pat.hir_id);
+        debug!("check_decl_local pat_ty {:?}", pat_ty);
         if pat_ty.references_error() {
```

Comment:
De-duplicate this please with a new method (including the `if`) (also to reduce diagnostic logic in happy path).